File: market-data/crypto-news/examples/swift/CryptoNews.swift
/**
 * Free Crypto News - Swift Examples
 * 
 * Comprehensive examples for iOS, macOS, watchOS, and tvOS.
 * 
 * Requirements:
 *   - Swift 5.9+
 *   - iOS 15+ / macOS 12+ / watchOS 8+ / tvOS 15+
 */

import Foundation

// MARK: - Models

struct NewsArticle: Codable, Identifiable {
    let id: String?
    let title: String
    let link: String
    let source: String
    let description: String?
    let timeAgo: String?
    let pubDate: Date?
    let tickers: [String]?
}

struct NewsResponse: Codable {
    let data: [NewsArticle]
    let total: Int
    let success: Bool
}

struct FearGreedData: Codable {
    let value: Int
    let classification: String
    let timestamp: Date?
}

struct SentimentData: Codable {
    let asset: String
    let label: String
    let score: Double
    let confidence: Double
    let articlesAnalyzed: Int?
}

struct TrendingTopic: Codable, Identifiable {
    var id: String { topic }
    let topic: String
    let count: Int
    let sentiment: String?
    let change: Double?
}

struct WhaleAlert: Codable, Identifiable {
    var id: String { "\(symbol)-\(usdValue)-\(timeAgo ?? "")" }
    let symbol: String
    let amount: Double
    let usdValue: Double
    let fromLabel: String
    let toLabel: String
    let type: String
    let blockchain: String
    let timeAgo: String?
}

struct TradingSignal: Codable, Identifiable {
    var id: String { "\(symbol)-\(action)-\(price)" }
    let symbol: String
    let action: String
    let price: Double
    let confidence: Double
    let reason: String?
    let targets: SignalTargets?
}

struct SignalTargets: Codable {
    let takeProfit: Double
    let stopLoss: Double
}

struct CoinPrice: Codable, Identifiable {
    var id: String { symbol }
    let symbol: String
    let name: String
    let price: Double
    let change24h: Double
    let marketCap: Double
    let volume24h: Double
}

// MARK: - API Client

actor CryptoNewsClient {
    private let baseURL: URL
    private let session: URLSession
    private let decoder: JSONDecoder
    
    init(apiKey: String? = nil, baseURL: String = "https://cryptocurrency.cv") {
        self.baseURL = URL(string: baseURL)!
        
        let config = URLSessionConfiguration.default
        config.httpAdditionalHeaders = [
            "User-Agent": "FCN-Swift/1.0"
        ]
        if let apiKey = apiKey {
            config.httpAdditionalHeaders?["Authorization"] = "Bearer \(apiKey)"
        }
        
        self.session = URLSession(configuration: config)
        
        self.decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .iso8601
    }
    
    // MARK: - News Endpoints
    
    func getNews(limit: Int = 20, source: String? = nil) async throws -> [NewsArticle] {
        var components = URLComponents(url: baseURL.appendingPathComponent("/api/news"), resolvingAgainstBaseURL: true)!
        components.queryItems = [URLQueryItem(name: "limit", value: "\(limit)")]
        if let source = source {
            components.queryItems?.append(URLQueryItem(name: "source", value: source))
        }
        
        let (data, _) = try await session.data(from: components.url!)
        let response = try decoder.decode(NewsResponse.self, from: data)
        return response.data
    }
    
    func searchNews(query: String, limit: Int = 20) async throws -> [NewsArticle] {
        var components = URLComponents(url: baseURL.appendingPathComponent("/api/search"), resolvingAgainstBaseURL: true)!
        components.queryItems = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "limit", value: "\(limit)")
        ]
        
        let (data, _) = try await session.data(from: components.url!)
        let response = try decoder.decode(NewsResponse.self, from: data)
        return response.data
    }
    
    func getBitcoinNews(limit: Int = 20) async throws -> [NewsArticle] {
        var components = URLComponents(url: baseURL.appendingPathComponent("/api/bitcoin"), resolvingAgainstBaseURL: true)!
        components.queryItems = [URLQueryItem(name: "limit", value: "\(limit)")]
        
        let (data, _) = try await session.data(from: components.url!)
        let response = try decoder.decode(NewsResponse.self, from: data)
        return response.data
    }
    
    func getBreakingNews() async throws -> [NewsArticle] {
        let url = baseURL.appendingPathComponent("/api/breaking")
        let (data, _) = try await session.data(from: url)
        let response = try decoder.decode(NewsResponse.self, from: data)
        return response.data
    }
    
    func getTrending(hours: Int = 24) async throws -> [TrendingTopic] {
        var components = URLComponents(url: baseURL.appendingPathComponent("/api/trending"), resolvingAgainstBaseURL: true)!
        components.queryItems = [URLQueryItem(name: "hours", value: "\(hours)")]
        
        let (data, _) = try await session.data(from: components.url!)
        let response = try decoder.decode([String: [TrendingTopic]].self, from: data)
        return response["data"] ?? []
    }
    
    // MARK: - Market Endpoints
    
    func getFearGreed() async throws -> FearGreedData {
        let url = baseURL.appendingPathComponent("/api/market/fear-greed")
        let (data, _) = try await session.data(from: url)
        let response = try decoder.decode([String: FearGreedData].self, from: data)
        return response["data"]!
    }
    
    func getPrices(symbols: [String]) async throws -> [CoinPrice] {
        var components = URLComponents(url: baseURL.appendingPathComponent("/api/prices"), resolvingAgainstBaseURL: true)!
        components.queryItems = [URLQueryItem(name: "symbols", value: symbols.joined(separator: ","))]
        
        let (data, _) = try await session.data(from: components.url!)
        let response = try decoder.decode([String: [CoinPrice]].self, from: data)
        return response["data"] ?? []
    }
    
    // MARK: - AI Endpoints
    
    func getSentiment(asset: String) async throws -> SentimentData {
        var components = URLComponents(url: baseURL.appendingPathComponent("/api/ai/sentiment"), resolvingAgainstBaseURL: true)!
        components.queryItems = [URLQueryItem(name: "asset", value: asset)]
        
        let (data, _) = try await session.data(from: components.url!)
        return try decoder.decode(SentimentData.self, from: data)
    }
    
    func getDigest() async throws -> String {
        let url = baseURL.appendingPathComponent("/api/ai/digest")
        let (data, _) = try await session.data(from: url)
        let response = try decoder.decode([String: String].self, from: data)
        return response["digest"] ?? ""
    }
    
    func ask(question: String) async throws -> String {
        var components = URLComponents(url: baseURL.appendingPathComponent("/api/ai/ask"), resolvingAgainstBaseURL: true)!
        components.queryItems = [URLQueryItem(name: "q", value: question)]
        
        let (data, _) = try await session.data(from: components.url!)
        let response = try decoder.decode([String: String].self, from: data)
        return response["answer"] ?? ""
    }
    
    // MARK: - Trading Endpoints
    
    func getSignals(limit: Int = 10) async throws -> [TradingSignal] {
        var components = URLComponents(url: baseURL.appendingPathComponent("/api/trading/signals"), resolvingAgainstBaseURL: true)!
        components.queryItems = [URLQueryItem(name: "limit", value: "\(limit)")]
        
        let (data, _) = try await session.data(from: components.url!)
        let response = try decoder.decode([String: [TradingSignal]].self, from: data)
        return response["data"] ?? []
    }
    
    func getWhaleAlerts(limit: Int = 20) async throws -> [WhaleAlert] {
        var components = URLComponents(url: baseURL.appendingPathComponent("/api/whales"), resolvingAgainstBaseURL: true)!
        components.queryItems = [URLQueryItem(name: "limit", value: "\(limit)")]
        
        let (data, _) = try await session.data(from: components.url!)
        let response = try decoder.decode([String: [WhaleAlert]].self, from: data)
        return response["data"] ?? []
    }
}

// MARK: - WebSocket Streaming

class CryptoNewsStream: NSObject, URLSessionWebSocketDelegate {
    private var webSocket: URLSessionWebSocketTask?
    private let baseURL: String
    
    var onNews: ((NewsArticle) -> Void)?
    var onPrice: ((CoinPrice) -> Void)?
    var onWhale: ((WhaleAlert) -> Void)?
    var onError: ((Error) -> Void)?
    var onConnect: (() -> Void)?
    var onDisconnect: (() -> Void)?
    
    init(baseURL: String = "wss://cryptocurrency.cv") {
        self.baseURL = baseURL
        super.init()
    }
    
    func connect() {
        let url = URL(string: "\(baseURL)/api/ws")!
        let session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
        webSocket = session.webSocketTask(with: url)
        webSocket?.resume()
        receiveMessage()
    }
    
    func subscribe(to channel: String) {
        let message = """
        {"action": "subscribe", "channel": "\(channel)"}
        """
        webSocket?.send(.string(message)) { [weak self] error in
            if let error = error {
                self?.onError?(error)
            }
        }
    }
    
    private func receiveMessage() {
        webSocket?.receive { [weak self] result in
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    self?.handleMessage(text)
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8) {
                        self?.handleMessage(text)
                    }
                @unknown default:
                    break
                }
                self?.receiveMessage()
                
            case .failure(let error):
                self?.onError?(error)
            }
        }
    }
    
    private func handleMessage(_ text: String) {
        guard let data = text.data(using: .utf8) else { return }
        
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        
        do {
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            guard let type = json?["type"] as? String,
                  let payloadData = json?["data"] else { return }
            
            let payload = try JSONSerialization.data(withJSONObject: payloadData)
            
            switch type {
            case "news":
                let article = try decoder.decode(NewsArticle.self, from: payload)
                DispatchQueue.main.async { self.onNews?(article) }
            case "price":
                let price = try decoder.decode(CoinPrice.self, from: payload)
                DispatchQueue.main.async { self.onPrice?(price) }
            case "whale":
                let whale = try decoder.decode(WhaleAlert.self, from: payload)
                DispatchQueue.main.async { self.onWhale?(whale) }
            default:
                break
            }
        } catch {
            onError?(error)
        }
    }
    
    func disconnect() {
        webSocket?.cancel(with: .normalClosure, reason: nil)
    }
    
    // MARK: - URLSessionWebSocketDelegate
    
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
        DispatchQueue.main.async { self.onConnect?() }
    }
    
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        DispatchQueue.main.async { self.onDisconnect?() }
    }
}

// MARK: - SwiftUI Views (iOS/macOS)

#if canImport(SwiftUI)
import SwiftUI

@MainActor
class NewsViewModel: ObservableObject {
    @Published var articles: [NewsArticle] = []
    @Published var trending: [TrendingTopic] = []
    @Published var fearGreed: FearGreedData?
    @Published var isLoading = false
    @Published var error: String?
    
    private let client = CryptoNewsClient()
    
    func loadNews() async {
        isLoading = true
        error = nil
        
        do {
            async let newsTask = client.getNews(limit: 20)
            async let trendingTask = client.getTrending()
            async let fearGreedTask = client.getFearGreed()
            
            let (news, trending, fearGreed) = try await (newsTask, trendingTask, fearGreedTask)
            
            self.articles = news
            self.trending = trending
            self.fearGreed = fearGreed
        } catch {
            self.error = error.localizedDescription
        }
        
        isLoading = false
    }
}

struct NewsListView: View {
    @StateObject private var viewModel = NewsViewModel()
    
    var body: some View {
        NavigationStack {
            Group {
                if viewModel.isLoading {
                    ProgressView("Loading...")
                } else if let error = viewModel.error {
                    VStack {
                        Text("Error: \(error)")
                            .foregroundColor(.red)
                        Button("Retry") {
                            Task { await viewModel.loadNews() }
                        }
                    }
                } else {
                    List {
                        // Fear & Greed Section
                        if let fg = viewModel.fearGreed {
                            Section("Market Sentiment") {
                                HStack {
                                    Text("Fear & Greed")
                                    Spacer()
                                    Text("\(fg.value) - \(fg.classification)")
                                        .foregroundColor(fg.value > 50 ? .green : .red)
                                }
                            }
                        }
                        
                        // Trending Section
                        if !viewModel.trending.isEmpty {
                            Section("Trending") {
                                ScrollView(.horizontal, showsIndicators: false) {
                                    HStack {
                                        ForEach(viewModel.trending.prefix(5)) { topic in
                                            Text(topic.topic)
                                                .padding(.horizontal, 12)
                                                .padding(.vertical, 6)
                                                .background(Color.blue.opacity(0.2))
                                                .cornerRadius(16)
                                        }
                                    }
                                }
                            }
                        }
                        
                        // News Section
                        Section("Latest News") {
                            ForEach(viewModel.articles) { article in
                                VStack(alignment: .leading, spacing: 4) {
                                    Text(article.title)
                                        .font(.headline)
                                    HStack {
                                        Text(article.source)
                                            .font(.caption)
                                            .foregroundColor(.secondary)
                                        if let timeAgo = article.timeAgo {
                                            Text("•")
                                                .foregroundColor(.secondary)
                                            Text(timeAgo)
                                                .font(.caption)
                                                .foregroundColor(.secondary)
                                        }
                                    }
                                }
                                .padding(.vertical, 4)
                            }
                        }
                    }
                    .refreshable {
                        await viewModel.loadNews()
                    }
                }
            }
            .navigationTitle("📰 Crypto News")
        }
        .task {
            await viewModel.loadNews()
        }
    }
}

#Preview {
    NewsListView()
}
#endif

// MARK: - Example Usage

@main
struct CryptoNewsApp {
    static func main() async {
        print("═══════════════════════════════════════════════════════")
        print("       Free Crypto News - Swift Examples")
        print("═══════════════════════════════════════════════════════\n")
        
        let client = CryptoNewsClient()
        
        // News
        print("📰 Latest News:")
        print("─────────────────────────────────────────────────────────")
        do {
            let news = try await client.getNews(limit: 5)
            for article in news {
                print("  • \(article.title)")
                print("    \(article.source) | \(article.timeAgo ?? "")")
            }
        } catch {
            print("  Error: \(error)")
        }
        
        // Fear & Greed
        print("\n😰 Fear & Greed Index:")
        print("─────────────────────────────────────────────────────────")
        do {
            let fg = try await client.getFearGreed()
            print("  Value: \(fg.value)")
            print("  Classification: \(fg.classification)")
        } catch {
            print("  Error: \(error)")
        }
        
        // Sentiment
        print("\n🎯 Bitcoin Sentiment:")
        print("─────────────────────────────────────────────────────────")
        do {
            let sentiment = try await client.getSentiment(asset: "BTC")
            print("  Label: \(sentiment.label)")
            print("  Score: \(String(format: "%.2f", sentiment.score))")
        } catch {
            print("  Error: \(error)")
        }
        
        // Trending
        print("\n🔥 Trending Topics:")
        print("─────────────────────────────────────────────────────────")
        do {
            let trending = try await client.getTrending()
            for topic in trending.prefix(5) {
                let emoji = topic.sentiment == "bullish" ? "🟢" :
                           topic.sentiment == "bearish" ? "🔴" : "⚪"
                print("  \(emoji) \(topic.topic) (\(topic.count) mentions)")
            }
        } catch {
            print("  Error: \(error)")
        }
        
        // Trading Signals
        print("\n📊 Trading Signals:")
        print("─────────────────────────────────────────────────────────")
        do {
            let signals = try await client.getSignals(limit: 3)
            for signal in signals {
                let emoji = signal.action == "buy" || signal.action == "long" ? "🟢" : "🔴"
                print("  \(emoji) \(signal.action.uppercased()) \(signal.symbol) @ $\(String(format: "%.2f", signal.price))")
            }
        } catch {
            print("  Error: \(error)")
        }
        
        // Whale Alerts
        print("\n🐋 Whale Alerts:")
        print("─────────────────────────────────────────────────────────")
        do {
            let whales = try await client.getWhaleAlerts(limit: 5)
            for whale in whales {
                let emoji = whale.type == "exchange_inflow" ? "📥" :
                           whale.type == "exchange_outflow" ? "📤" : "🔄"
                print("  \(emoji) \(whale.symbol): $\(Int(whale.usdValue).formatted())")
            }
        } catch {
            print("  Error: \(error)")
        }
        
        print("\n═══════════════════════════════════════════════════════")
        print("       Examples Complete!")
        print("═══════════════════════════════════════════════════════")
    }
}

File: market-data/crypto-news/examples/kotlin/CryptoNews.kt
/**
 * Free Crypto News - Kotlin Examples
 * 
 * Comprehensive examples for Android, JVM, and Kotlin Multiplatform.
 * 
 * Dependencies:
 *   implementation("io.ktor:ktor-client-core:2.3.0")
 *   implementation("io.ktor:ktor-client-cio:2.3.0")
 *   implementation("io.ktor:ktor-client-content-negotiation:2.3.0")
 *   implementation("io.ktor:ktor-serialization-kotlinx-json:2.3.0")
 *   implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.7.0")
 */

package com.freecryptonews

import kotlinx.coroutines.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.websocket.*
import io.ktor.client.request.*
import io.ktor.serialization.kotlinx.json.*
import io.ktor.websocket.*

// ═══════════════════════════════════════════════════════════════
// Models
// ═══════════════════════════════════════════════════════════════

@Serializable
data class NewsArticle(
    val title: String,
    val link: String,
    val source: String,
    val description: String? = null,
    val timeAgo: String? = null,
    val pubDate: String? = null,
    val tickers: List<String>? = null
)

@Serializable
data class NewsResponse(
    val data: List<NewsArticle>,
    val total: Int,
    val success: Boolean
)

@Serializable
data class FearGreedData(
    val value: Int,
    val classification: String,
    val timestamp: String? = null
)

@Serializable
data class FearGreedResponse(
    val data: FearGreedData
)

@Serializable
data class SentimentData(
    val asset: String,
    val label: String,
    val score: Double,
    val confidence: Double,
    val articlesAnalyzed: Int? = null
)

@Serializable
data class TrendingTopic(
    val topic: String,
    val count: Int,
    val sentiment: String? = null,
    val change: Double? = null
)

@Serializable
data class TrendingResponse(
    val data: List<TrendingTopic>
)

@Serializable
data class WhaleAlert(
    val symbol: String,
    val amount: Double,
    val usdValue: Double,
    val fromLabel: String,
    val toLabel: String,
    val type: String,
    val blockchain: String,
    val timeAgo: String? = null
)

@Serializable
data class WhalesResponse(
    val data: List<WhaleAlert>
)

@Serializable
data class TradingSignal(
    val symbol: String,
    val action: String,
    val price: Double,
    val confidence: Double,
    val reason: String? = null,
    val targets: SignalTargets? = null
)

@Serializable
data class SignalTargets(
    val takeProfit: Double,
    val stopLoss: Double
)

@Serializable
data class SignalsResponse(
    val data: List<TradingSignal>
)

@Serializable
data class CoinPrice(
    val symbol: String,
    val name: String,
    val price: Double,
    val change24h: Double,
    val marketCap: Double,
    val volume24h: Double
)

// ═══════════════════════════════════════════════════════════════
// Client
// ═══════════════════════════════════════════════════════════════

class CryptoNewsClient(
    private val apiKey: String? = null,
    private val baseUrl: String = "https://cryptocurrency.cv"
) {
    private val json = Json { 
        ignoreUnknownKeys = true 
        isLenient = true
    }
    
    private val client = HttpClient(CIO) {
        install(ContentNegotiation) {
            json(json)
        }
        install(WebSockets)
    }

    // ─────────────────────────────────────────────────────────────
    // News Endpoints
    // ─────────────────────────────────────────────────────────────

    suspend fun getNews(limit: Int = 20, source: String? = null): List<NewsArticle> {
        val response: NewsResponse = client.get("$baseUrl/api/news") {
            parameter("limit", limit)
            source?.let { parameter("source", it) }
            apiKey?.let { header("Authorization", "Bearer $it") }
        }.body()
        return response.data
    }

    suspend fun searchNews(query: String, limit: Int = 20): List<NewsArticle> {
        val response: NewsResponse = client.get("$baseUrl/api/search") {
            parameter("q", query)
            parameter("limit", limit)
        }.body()
        return response.data
    }

    suspend fun getBitcoinNews(limit: Int = 20): List<NewsArticle> {
        val response: NewsResponse = client.get("$baseUrl/api/bitcoin") {
            parameter("limit", limit)
        }.body()
        return response.data
    }

    suspend fun getBreakingNews(): List<NewsArticle> {
        val response: NewsResponse = client.get("$baseUrl/api/breaking").body()
        return response.data
    }

    suspend fun getTrending(hours: Int = 24): List<TrendingTopic> {
        val response: TrendingResponse = client.get("$baseUrl/api/trending") {
            parameter("hours", hours)
        }.body()
        return response.data
    }

    // ─────────────────────────────────────────────────────────────
    // Market Endpoints
    // ─────────────────────────────────────────────────────────────

    suspend fun getFearGreed(): FearGreedData {
        val response: FearGreedResponse = client.get("$baseUrl/api/market/fear-greed").body()
        return response.data
    }

    suspend fun getPrices(symbols: List<String>): List<CoinPrice> {
        val response: Map<String, List<CoinPrice>> = client.get("$baseUrl/api/prices") {
            parameter("symbols", symbols.joinToString(","))
        }.body()
        return response["data"] ?: emptyList()
    }

    // ─────────────────────────────────────────────────────────────
    // AI Endpoints
    // ─────────────────────────────────────────────────────────────

    suspend fun getSentiment(asset: String): SentimentData {
        return client.get("$baseUrl/api/ai/sentiment") {
            parameter("asset", asset)
        }.body()
    }

    suspend fun getDigest(): String {
        val response: Map<String, String> = client.get("$baseUrl/api/ai/digest").body()
        return response["digest"] ?: ""
    }

    suspend fun ask(question: String): String {
        val response: Map<String, String> = client.get("$baseUrl/api/ai/ask") {
            parameter("q", question)
        }.body()
        return response["answer"] ?: ""
    }

    // ─────────────────────────────────────────────────────────────
    // Trading Endpoints
    // ─────────────────────────────────────────────────────────────

    suspend fun getSignals(limit: Int = 10): List<TradingSignal> {
        val response: SignalsResponse = client.get("$baseUrl/api/trading/signals") {
            parameter("limit", limit)
        }.body()
        return response.data
    }

    suspend fun getWhaleAlerts(limit: Int = 20): List<WhaleAlert> {
        val response: WhalesResponse = client.get("$baseUrl/api/whales") {
            parameter("limit", limit)
        }.body()
        return response.data
    }

    fun close() {
        client.close()
    }
}

// ═══════════════════════════════════════════════════════════════
// WebSocket Streaming
// ═══════════════════════════════════════════════════════════════

class CryptoNewsStream(
    private val baseUrl: String = "wss://cryptocurrency.cv"
) {
    private val client = HttpClient(CIO) {
        install(WebSockets)
    }
    
    private val json = Json { ignoreUnknownKeys = true }
    
    var onNews: ((NewsArticle) -> Unit)? = null
    var onPrice: ((CoinPrice) -> Unit)? = null
    var onWhale: ((WhaleAlert) -> Unit)? = null
    var onError: ((Throwable) -> Unit)? = null
    var onConnect: (() -> Unit)? = null
    var onDisconnect: (() -> Unit)? = null

    suspend fun connect(channels: List<String> = listOf("news")) {
        try {
            client.webSocket("$baseUrl/api/ws") {
                onConnect?.invoke()
                
                // Subscribe to channels
                for (channel in channels) {
                    send("""{"action": "subscribe", "channel": "$channel"}""")
                }
                
                // Receive messages
                for (frame in incoming) {
                    when (frame) {
                        is Frame.Text -> handleMessage(frame.readText())
                        else -> {}
                    }
                }
            }
        } catch (e: Exception) {
            onError?.invoke(e)
        } finally {
            onDisconnect?.invoke()
        }
    }

    private fun handleMessage(text: String) {
        try {
            val element = json.parseToJsonElement(text).jsonObject
            val type = element["type"]?.jsonPrimitive?.content
            val data = element["data"]

            when (type) {
                "news" -> {
                    data?.let {
                        val article = json.decodeFromJsonElement<NewsArticle>(it)
                        onNews?.invoke(article)
                    }
                }
                "price" -> {
                    data?.let {
                        val price = json.decodeFromJsonElement<CoinPrice>(it)
                        onPrice?.invoke(price)
                    }
                }
                "whale" -> {
                    data?.let {
                        val whale = json.decodeFromJsonElement<WhaleAlert>(it)
                        onWhale?.invoke(whale)
                    }
                }
            }
        } catch (e: Exception) {
            onError?.invoke(e)
        }
    }

    fun close() {
        client.close()
    }
}

// ═══════════════════════════════════════════════════════════════
// Android ViewModel Example
// ═══════════════════════════════════════════════════════════════

/*
// For Android, use with Jetpack Compose:

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class NewsViewModel : ViewModel() {
    private val client = CryptoNewsClient()
    
    private val _articles = MutableStateFlow<List<NewsArticle>>(emptyList())
    val articles: StateFlow<List<NewsArticle>> = _articles
    
    private val _fearGreed = MutableStateFlow<FearGreedData?>(null)
    val fearGreed: StateFlow<FearGreedData?> = _fearGreed
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading
    
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    fun loadNews() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null
            
            try {
                _articles.value = client.getNews(20)
                _fearGreed.value = client.getFearGreed()
            } catch (e: Exception) {
                _error.value = e.message
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    override fun onCleared() {
        super.onCleared()
        client.close()
    }
}

// Composable:
@Composable
fun NewsScreen(viewModel: NewsViewModel = viewModel()) {
    val articles by viewModel.articles.collectAsState()
    val isLoading by viewModel.isLoading.collectAsState()
    
    LaunchedEffect(Unit) {
        viewModel.loadNews()
    }
    
    if (isLoading) {
        CircularProgressIndicator()
    } else {
        LazyColumn {
            items(articles) { article ->
                Text(article.title)
            }
        }
    }
}
*/

// ═══════════════════════════════════════════════════════════════
// Examples
// ═══════════════════════════════════════════════════════════════

suspend fun main() {
    println("═══════════════════════════════════════════════════════")
    println("       Free Crypto News - Kotlin Examples")
    println("═══════════════════════════════════════════════════════\n")

    val client = CryptoNewsClient()

    try {
        // News
        println("📰 Latest News:")
        println("─────────────────────────────────────────────────────────")
        val news = client.getNews(5)
        news.forEach { article ->
            println("  • ${article.title}")
            println("    ${article.source} | ${article.timeAgo ?: ""}")
        }

        // Fear & Greed
        println("\n😰 Fear & Greed Index:")
        println("─────────────────────────────────────────────────────────")
        val fg = client.getFearGreed()
        println("  Value: ${fg.value}")
        println("  Classification: ${fg.classification}")

        // Sentiment
        println("\n🎯 Bitcoin Sentiment:")
        println("─────────────────────────────────────────────────────────")
        val sentiment = client.getSentiment("BTC")
        println("  Label: ${sentiment.label}")
        println("  Score: ${"%.2f".format(sentiment.score)}")

        // Trending
        println("\n🔥 Trending Topics:")
        println("─────────────────────────────────────────────────────────")
        val trending = client.getTrending()
        trending.take(5).forEach { topic ->
            val emoji = when (topic.sentiment) {
                "bullish" -> "🟢"
                "bearish" -> "🔴"
                else -> "⚪"
            }
            println("  $emoji ${topic.topic} (${topic.count} mentions)")
        }

        // Trading Signals
        println("\n📊 Trading Signals:")
        println("─────────────────────────────────────────────────────────")
        val signals = client.getSignals(3)
        signals.forEach { signal ->
            val emoji = if (signal.action in listOf("buy", "long")) "🟢" else "🔴"
            println("  $emoji ${signal.action.uppercase()} ${signal.symbol} @ $${String.format("%.2f", signal.price)}")
        }

        // Whale Alerts
        println("\n🐋 Whale Alerts:")
        println("─────────────────────────────────────────────────────────")
        val whales = client.getWhaleAlerts(5)
        whales.forEach { whale ->
            val emoji = when (whale.type) {
                "exchange_inflow" -> "📥"
                "exchange_outflow" -> "📤"
                else -> "🔄"
            }
            println("  $emoji ${whale.symbol}: $${"%,.0f".format(whale.usdValue)}")
        }

        // Search
        println("\n🔍 Search 'ethereum ETF':")
        println("─────────────────────────────────────────────────────────")
        val searchResults = client.searchNews("ethereum ETF", 3)
        searchResults.forEach { article ->
            println("  • ${article.title}")
        }

    } catch (e: Exception) {
        println("Error: ${e.message}")
    } finally {
        client.close()
    }

    println("\n═══════════════════════════════════════════════════════")
    println("       Examples Complete!")
    println("═══════════════════════════════════════════════════════")
}
